test/autograd/test_fallback.py

```diff
@@ -0,0 +1,302 @@
+        y = op(x).sum()
+        with self._check_ctx(mode):
+            y.backward()
+            self.assertEqual(x.grad, x.cos())
+
+    @parametrize("mode", ("nothing", "warn"))
+    def test_inplace_autograd_function_registered_to_cpu(self, mode):
+        lib = Library(self.test_ns, "FRAGMENT")
+        lib.define("foo(Tensor(a!) self) -> Tensor(a!)")
+
+        class NumpySin_(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                ctx.save_for_backward(x.clone())
+                x_np = x.detach().numpy()
+                np.sin(x_np, out=x_np)
+                ctx.mark_dirty(x)
+                return x
+
+            @staticmethod
+            def backward(ctx, gx):
+                x, = ctx.saved_tensors
+                return gx * x.cos()
+
+        lib.impl("foo", NumpySin_.apply, "CPU")
+        op = self.get_op("foo")
+
+        x = torch.randn(3, requires_grad=True)
+        z = x.clone()
+        w = z[0]
+        y = op(w)
+
+        expected = torch.zeros_like(x)
+        expected[0] = x[0].cos()
+        with self._check_ctx(mode):
+            gx, = torch.autograd.grad(y, x, torch.ones_like(y), retain_graph=True)
+            self.assertEqual(gx, expected)
+
+        expected = torch.ones_like(x)
+        expected[0] = x[0].cos()
+        with self._check_ctx(mode):
+            gx, = torch.autograd.grad(z, x, torch.ones_like(z))
+            self.assertEqual(gx, expected)
+
+    @parametrize("mode", ("nothing", "warn"))
+    def test_post_autograd_returns_leaf(self, mode):
+        lib = self.get_lib()
+        lib.define("foo(Tensor a) -> (Tensor, Tensor)")
+        op = self.get_op("foo")
+
+        lib.impl("foo", lambda a: (a.clone(), a.clone().detach().requires_grad_()), "CPU")
+        x = torch.randn(3, requires_grad=True)
+        y, z = op(x)
+        with self._check_ctx(mode):
+            z.sum().backward()
+
+    @parametrize("mode", ("nothing", "warn"))
+    def test_post_autograd_returns_mix_of_requires_grad_tensors(self, mode):
+        lib = self.get_lib()
+        lib.define("foo(Tensor a, Tensor b) -> (Tensor, Tensor, Tensor)")
+        op = self.get_op("foo")
+
+        def foo_impl(a, b):
+            with torch.no_grad():
+                x = a.clone()
+                z = b.clone()
+            y = a * b
+            return x, y, z
+
+        lib.impl("foo", foo_impl, "CPU")
+        a = torch.randn(3, requires_grad=True)
+        b = torch.randn(3, requires_grad=True)
+        x, y, z = op(a, b)
+
+        with self._check_ctx(mode):
+            torch.autograd.grad(x, (a, b), torch.ones_like(x), allow_unused=True, retain_graph=True)
+
+        with self._check_ctx(mode):
+            torch.autograd.grad(y, (a, b), torch.ones_like(y), allow_unused=True, retain_graph=True)
+
+        with self._check_ctx(mode):
+            torch.autograd.grad(z, (a, b), torch.ones_like(z), allow_unused=True, retain_graph=True)
```

Comment:
Same here (see below), I'd expect the `x` and `z` ones to error in the mode="nothing" case.